### PR TITLE
quic: refactor QuicPersistentInfoImpl

### PR DESCRIPTION
**Pull request by danzh2010**
Commit Message: Refactor QuicPersistentInfoImpl. Move out the fields that are not persistent across cluster, i.e. transport_socket_factory_ and client_conext_, and crypto_config_.

This is to prepare for changing QuicPersistentInfoImpl ownership to cluster in https://github.com/envoyproxy/envoy/pull/19862.

Risk Level: low, refactoring only
Testing: existing tests pass
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A
Part of https://github.com/envoyproxy/envoy/issues/19861, https://github.com/envoyproxy/envoy/issues/18715

### CODE REVIEW

**Comment by danzh2010** [comment]
@RyanTheOptimist This is good for review!

**Comment by RyanTheOptimist** [comment]
Now that #19860 has been merged, can you update this PR to merge in main?

**Comment by danzh2010** [comment]
> Now that #19860 has been merged, can you update this PR to merge in main?

CI cleared. PTAL!

**Review by RyanTheOptimist — DISMISSED** [review]
Thanks for rebasing! So much easier to review. This looks good modulo just a few nits.

File: source/common/http/http3/conn_pool.cc [hunk 1]
```diff
@@ -55,6 +56,14 @@ void ActiveClient::onMaxStreamsChanged(uint32_t num_streams) {
   }
 }
 
+const Envoy::Ssl::ClientContextConfig&
+getConfig(Network::TransportSocketFactory& transport_socket_factory) {
+  auto* quic_socket_factory =
+      dynamic_cast<Quic::QuicClientTransportSocketFactory*>(&transport_socket_factory);
+  ASSERT(quic_socket_factory != nullptr);
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/common/http/http3/conn_pool.cc]: done

File: source/common/http/http3/conn_pool.cc [hunk 2]
```diff
@@ -99,13 +95,52 @@ void Http3ConnPoolImpl::onConnected(Envoy::ConnectionPool::ActiveClient&) {
 // Make sure all connections are torn down before quic_info_ is deleted.
 Http3ConnPoolImpl::~Http3ConnPoolImpl() { destructAllConnections(); }
 
+std::shared_ptr<quic::QuicCryptoClientConfig> Http3ConnPoolImpl::cryptoConfig() {
+  auto* quic_socket_factory =
+      dynamic_cast<Quic::QuicClientTransportSocketFactory*>(&host_->transportSocketFactory());
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/common/http/http3/conn_pool.cc]: done

File: source/common/http/http3/conn_pool.cc [hunk 3]
```diff
@@ -99,13 +95,52 @@ void Http3ConnPoolImpl::onConnected(Envoy::ConnectionPool::ActiveClient&) {
...
+  ASSERT(quic_socket_factory != nullptr);
+  auto context = quic_socket_factory->sslCtx();
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/common/http/http3/conn_pool.cc]: done

File: source/common/quic/client_connection_factory_impl.h [hunk 4]
```diff
@@ -17,28 +17,10 @@ namespace Quic {
...
 // Information which can be shared across connections, though not across threads.
 struct PersistentQuicInfoImpl : public Http::PersistentQuicInfo {
-  PersistentQuicInfoImpl(Event::Dispatcher& dispatcher,
-                         Network::TransportSocketFactory& transport_socket_factory,
-                         TimeSource& time_source, uint32_t remote_port,
-                         const quic::QuicConfig& quic_config, uint32_t buffer_limit);
-
-  // Returns the most recent crypto config from transport_socket_factory_;
-  std::shared_ptr<quic::QuicCryptoClientConfig> cryptoConfig();
+  PersistentQuicInfoImpl(Event::Dispatcher& dispatcher, uint32_t buffer_limit);
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/common/quic/client_connection_factory_impl.h:20]: It's the connection send buffer (soft) limit. I modified the comment about buffer_limit_ member variable to be more specific.

**Review by RyanTheOptimist — APPROVED** [review]

**Comment by danzh2010** [comment]
@RyanTheOptimist can this be merged?

**Pull request merged**